File: T-Bank/Algorithms-and-Data-Structures-2025/1sprint-IntroductionToSorting-BinarySearch/contest/C-Отгадай_число/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(logN)
    Memory Complexity O(1)
    """
    n: int = int(sys.stdin.readline().rstrip())
    left: int = 1
    right: int = n
    while left <= right:
        middle: int = left + ((right - left) >> 1)
        sys.stdout.write(f'{middle}\n')
        sys.stdout.flush()
        sign: str = sys.stdin.readline().rstrip()
        if sign == '>=':
            left = middle + 1
        else:
            right = middle - 1
    sys.stdout.write(f'! {right}\n')


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/1sprint-IntroductionToSorting-BinarySearch/contest/C-Отгадай_число/solution.cpp
#include <iostream>
#include <string>


void solution() {
    /*
    Time Complexity O(logN)
    Memory Complexity O(1)
    */
    int n;
    std::cin >> n;
    int left = 1;
    int right = n;
    while (left <= right) {
        int middle = left + (right - left) / 2;
        std::cout << middle << "\n";
        std::fflush(stdout);
        std::string sign;
        std::cin >> sign;
        if (sign == ">=") left = middle + 1;
        else right = middle - 1;
    }
    std::printf("! %d\n", right);
}


int main() {
    solution();
}